### PR TITLE
Add entity API integration for field discovery

### PR DESCRIPTION
**Pull request by nedSaf**

### CODE REVIEW

File: plugins/restful/RestfulEntityBase.php [hunk 1]
```diff
@@ -1449,12 +1449,16 @@ public function getPublicFields() {
           // Set the column name.
           $info['column'] = key($field['columns']);
         }
+
+        if ($this->getMethod() == \RestfulInterface::OPTIONS) {
+          $info += $this->getFieldInfoAndFormSchema($field);
```
Comment (e0ipso) [inline plugins/restful/RestfulEntityBase.php:1133]: I'd rather not segregate this into a different method for Developer Experience. After all this is an optional feature.

I have to admit that I was tempted to go this route too, but I backed because it seemed cumbersome to implement. I imagine devs having to match keys in both functions and having to implement 2 different methods and I see potential bugs all over the place.

Do you have strong feelings about this? I can still be talked into it :stuck_out_tongue:

**Comment by e0ipso** [comment]
@nader77 I'm so sorry I did not see this before I jumped and replicate the same you have don here in 4659c2d.

Glad to see you in the issue queue!!!

**Comment by nedSaf** [comment]
@mateu-aguilo-bosch No problem, i will fix the conflicts.

We have to ask amitai regarding the route we took here :)

File: plugins/restful/RestfulEntityBase.php (hunk 1 above)
Comment (amitaibu) [inline plugins/restful/RestfulEntityBase.php:1133]: >  After all this is an optional feature.

My take is that by splitting to own function, you actually make it easier to override to your own needs. also getFieldInfoAndFormSchema() isn't too hard to follow, so I think it's fine as a separate method + it makes this `getPublicFields()` a bit more readable IMO.

Comment (e0ipso) [inline plugins/restful/RestfulEntityBase.php:1133]: Would you put in place some kind of checks to make sure both things match?

Comment (amitaibu) [inline plugins/restful/RestfulEntityBase.php:1133]: what do you mean?

File: plugins/restful/RestfulBase.php [hunk 2]
```diff
@@ -569,6 +568,7 @@ public function options($path = '', array $request = array()) {
         'type' => NULL,
         'read_only' => FALSE,
         'cardinality' => 1,
+        'required' => 0,
```
Comment (e0ipso) [inline plugins/restful/RestfulBase.php]: Let's make it FALSE.

File: plugins/restful/RestfulBase.php [hunk 3]
```diff
@@ -558,10 +558,9 @@ public function options($path = '', array $request = array()) {
 
     $output = array();
     // Default options for the discovery information.
-    // TODO: get this info using Entity API if it's an entity.
```
Comment (e0ipso) [inline plugins/restful/RestfulBase.php:561]: Can we keep this?

File: plugins/restful/RestfulBase.php [hunk 4]
```diff
@@ -558,10 +558,9 @@ public function options($path = '', array $request = array()) {
...
     $discovery_defaults = array(
       'info' => array(
-        'name' => '',
```
Comment (e0ipso) [inline plugins/restful/RestfulBase.php:564]: Why the change?

File: plugins/restful/RestfulBase.php (hunk 3 above)
Comment (amitaibu) [inline plugins/restful/RestfulBase.php:561]: but we are infact already do it in https://github.com/mateu-aguilo-bosch/restful/pull/2/files#diff-3c241a6c3b4e93a88a7da3e91c3e9358R1173

File: plugins/restful/RestfulBase.php (hunk 4 above)
Comment (amitaibu) [inline plugins/restful/RestfulBase.php:564]: `name` might be taken as `machine_name` -- for me `label` is more clear that it's the human readable name.

File: plugins/restful/RestfulEntityBase.php (hunk 1 above)
Comment (e0ipso) [inline plugins/restful/RestfulEntityBase.php:1133]: Do we check things like 'image' in the field definition and 'imgae' in the discovery options?

File: plugins/restful/RestfulBase.php (hunk 3 above)
Comment (e0ipso) [inline plugins/restful/RestfulBase.php:561]: True! Sorry.

File: plugins/restful/RestfulBase.php (hunk 4 above)
Comment (e0ipso) [inline plugins/restful/RestfulBase.php:564]: 👌

File: plugins/restful/RestfulBase.php (hunk 3 above)
Comment (amitaibu) [inline plugins/restful/RestfulBase.php:561]: we actually do very little there (not taking care of more special fields such as entity-reference), but that can be a follow up maybe?

File: plugins/restful/RestfulEntityBase.php (hunk 1 above)
Comment (amitaibu) [inline plugins/restful/RestfulEntityBase.php:1133]: In case of image/ file the type of the field (i.e. under data) would appear as `image`

File: plugins/restful/RestfulBase.php (hunk 2 above)
Comment (amitaibu) [inline plugins/restful/RestfulBase.php]: @nader77 can you change this please.

**Comment by nedSaf** [comment]
@mateu-aguilo-bosch @amitaibu 'required' changed.

**Comment by e0ipso** [comment]
Awesome, merging this. Can you open a new one to fix the tests that will now fail? Thanks!
